### DISCUSSION TITLE
How to reduce the latency of a select query?

### DISCUSSION THREAD

**Discussion by mrdbourke** [supabase/supabase#5532] [Questions] [answered]
Hi there,

I've got a relatively small Supabase project going (but I'd like to scale it up).

Right now I'm performing `select()` queries for single matching ID's but they seem to be performing slower than I'd like.

I'm new to JavaScript & Supabase but I thought for sure these queries would be running faster than they are now.

Right now it seems to be running at **~1-2 seconds (sometimes more) for returning a single row**.

Perhaps it's worth loading the whole database as JSON on page load and then querying that? (or is this silly? + doesn't scale)

My main question(s) are:
1. What's a reasonable performance time one should expect for a frontend to query a Supabase backend - ideally I'd like somewhere between 50-250ms, is this possible?
2. I am on the free tier - does this affect things? 

See the code: https://github.com/mrdbourke/food-data-central-database-creation-with-supabase
See the demo: https://food-data-central-database-creation-with-supabase.vercel.app/
Short video:

https://user-images.githubusercontent.com/16750345/154630133-6c865bfe-beef-4526-8d96-8d60e59e307c.mov

**Comment by steve-chavez** [comment, marked as answer]
Hello,

I think the main problem here is latency, queries should resolve pretty fast(few milliseconds) at the db side, even with millions of rows(assuming an index is present).

Latency depends on where you created your db and from where is the request coming from, the closer the better(less latency). In my case(south america), I get `~0.6 seconds` on average on your test app.

> Perhaps it's worth loading the whole database as JSON on page load and then querying that? (or is this silly? + doesn't scale)

If you need low latency, yes, that would be the best choice. I assume you don't really want the "whole db" to show in a dropdown though, you'd likely need to paginate results somehow in your UI.

> What's a reasonable performance time one should expect for a frontend to query a Supabase backend - ideally I'd like somewhere between 50-250ms, is this possible?

For the hosted service(behind Cloudflare), the response latency should be 350ms on average(assuming server/client on the same region).

> I am on the free tier - does this affect things?

For now you can only use a single region(both free and pro) but a pro offering for multi-region setups is planned.

**Reply by mrdbourke** [reply]
Thank you for this, this clarifies a bunch of things!

**Reply by steve-chavez** [reply]
@mrdbourke No problem, glad to help! Btw, there are numbers on latency here: https://status.supabase.com/

**Comment by Emre-Zero** [comment]
I noticed this issue as well when using Supabase database API inside Next.js/Vercel edge API function.
My API calls often take anywhere from 1 to 6 seconds! Not great.
I've also tried their experimental Edge functions with no improvement.
I emailed feedback to Vercel support.

I'm experimenting with using Cloudflare Worker to call Supabase database API instead.
It's actually faster than calling Supabase directly from client-side/browser.

Vercel edge function = 1-6s 🥲
Client side = 200ms
CF worker = ~80ms 🚀

I'm guessing this is because CF Worker has optimized network route to Supabase's AWS servers.

So you can essentially supercharge Supabase's database REST API by proxying it with a CF Worker. 

CF Worker API boilerplate: https://github.com/codewithkristian/serverless-api
You can use `import { createClient } from "@supabase/supabase-js";` library w/o issues inside CF Workers.

**Reply by romanrokon** [reply]
The marked answer didn't help you?
I think your latency issue is related to edge function's region.

**Reply by Emre-Zero** [reply]
@r0mankon I don't think it's a region issue. Maybe some kind of network issue however with how routing is done within Vercel's platform. I don't think it's an issue with Supabase.

**Reply by o1Suleyman** [reply]
> @r0mankon I don't think it's a region issue. Maybe some kind of network issue however with how routing is done within Vercel's platform. I don't think it's an issue with Supabase.

what do you mean routing is done within Vercel's platform?
